### PR TITLE
[BEAM-10887] Expose clear() to FnApiTimer.

### PR DESCRIPTION
**Pull request by boyuanzz**
This PR exposes `clear()` API to FnApiTimer, just like what we do for python SDK: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/runners/worker/bundle_processor.py#L661.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/Timer.java [hunk 1]
```diff
@@ -81,6 +81,9 @@
    */
   void setRelative();
 
+  /** Clears the timer. */
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/state/Timer.java:84]: This contract is vague.

**Review by lukecwik — CHANGES_REQUESTED** [review]
The contract is vague so if timer A and timer B are scheduled to fire as part of the same bundle and A clears B should B fire?

Also some runners gather timers and fire them as groups or fire them in timer loops.

We should add a bunch of ValidatesRunner tests covering:
* the various time domains
* clearing a timer that wasn't set
* clearing a timer that was set as part of the same bundle
* clearing a timer that was set as part of a different bundle

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/Timer.java (hunk 1 above)
Comment (boyuanzz) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/Timer.java:84]: There are 2 conditions that we know a cleared timer should never been fired:

- Timer(key = A) is set first then cleared within the same bundle.

- A never-set timer is cleared.

One condition that we know a cleared timer will be fired:

- Timer(key = A) is cleared first then set again.

And the contract is vague under certain conditions(e.g, the runner may or may not clear the timer depends on the implementation details of the runner):

- Timer(key = A)  is being scheduled by service meanwhile user code just clears Timer(key = A). This happens when set and clear calls happen on different bundles.

I'll add ValidatesRunner tests to cover these.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java [hunk 2]
```diff
@@ -1624,7 +1630,15 @@ public TimerUpdateBuilder setTimer(TimerData setTimer) {
...
-        setTimers.remove(deletedTimer);
+        TimerData timerToDelete = null;
+        for (TimerData timer : setTimers) {
+          if (timer.getDomain().equals(deletedTimer.getDomain())
+              && timer.getNamespace().equals(deletedTimer.getNamespace())
+              && timer.getTimerId().equals(deletedTimer.getTimerId())) {
+            timerToDelete = timer;
+          }
+        }
+        setTimers.remove(timerToDelete);
```
Comment (ibzib) [inline runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java:1641]: Should we warn the user if they try to clear an unset timer?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 3]
```diff
@@ -4609,6 +4609,346 @@ public void onTimer(
...
+                      .plus(Duration.standardMinutes(4)))
+              .advanceWatermarkToInfinity();
+
+      PCollection<Integer> output = pipeline.apply(stream).apply(ParDo.of(fn));
+      PAssert.that(output)
+          .satisfies(
+              (Iterable<Integer> input) -> {
+                assertEquals(1, Iterables.frequency(input, 3));
+                assertTrue(
+                    Iterables.frequency(input, 42) == 1 || Iterables.frequency(input, 42) == 2);
```
Comment (ibzib) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java:4815]: Can you add a comment explaining why this behavior is not deterministic?

**Review by ibzib — COMMENTED** [review]
Since this is a user-facing API change, it's worth adding a notification to CHANGES.md.

File: runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java (hunk 2 above)
Comment (boyuanzz) [reply runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java:1641]: Clearing an unset timer seems to be like a no-op to me.

**Review by ibzib — APPROVED** [review]
LGTM

**Pull request closed without merging**